core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -242,22 +243,39 @@ private boolean hasContentTypeOrCanAutoDetect(final RestRequest restRequest, fin
                 // be removed!
                 deprecationLogger.deprecated("Plain text request bodies are deprecated. Use request parameters or body " +
                     "in a supported format.");
-            } else if (isContentTypeRequired) {
-                return false;
-            } else {
-                deprecationLogger.deprecated("Content type detection for rest requests is deprecated. Specify the content type using " +
-                    "the [Content-Type] header.");
-                XContentType xContentType = XContentFactory.xContentType(restRequest.content());
-                if (xContentType == null) {
+            } else if (restHandler != null && restHandler.supportsContentStream() && restRequest.header("Content-Type") != null) {
+                final String lowercaseMediaType = restRequest.header("Content-Type").toLowerCase(Locale.ROOT);
+                // we also support line-delimited JSON, which isn't official and has a few variations
+                // http://specs.okfnlabs.org/ndjson/
+                // https://github.com/ndjson/ndjson-spec/blob/48ea03cea6796b614cfbff4d4eb921f0b1d35c26/specification.md
+                if (lowercaseMediaType.equals("application/x-ldjson") || lowercaseMediaType.equals("application/x-ndjson")) {
+                    restRequest.setXContentType(XContentType.JSON);
+                } else if (isContentTypeRequired) {
```

Comment:
shall we test, just to make sure, what happens with bulk and msearch when content-type is required and not provided? Wondering if all of these branches are covered by tests.